pkg/cloudprovider/vsphere/loadbalancer/lbprovider.go

```diff
@@ -0,0 +1,210 @@
+		return NewInMemoryNsxtBroker(nsxtSim.SimulatedIPPools...), nil
+	}
+
+	retriesConfig := nsxt.ClientRetriesConfiguration{
+		MaxRetries:      nsxtConfig.MaxRetries,
+		RetryMinDelay:   nsxtConfig.RetryMinDelay,
+		RetryMaxDelay:   nsxtConfig.RetryMaxDelay,
+		RetryOnStatuses: nsxtConfig.RetryOnStatusCodes,
+	}
+	cfg := nsxt.Configuration{
+		BasePath:             "/api/v1",
+		Host:                 nsxtConfig.Host,
+		Scheme:               "https",
+		UserAgent:            "nsxt-lb-provider/" + Version,
+		UserName:             nsxtConfig.User,
+		Password:             nsxtConfig.Password,
+		RemoteAuth:           nsxtConfig.RemoteAuth,
+		ClientAuthCertFile:   nsxtConfig.ClientAuthCertFile,
+		ClientAuthKeyFile:    nsxtConfig.ClientAuthKeyFile,
+		CAFile:               nsxtConfig.CAFile,
+		Insecure:             nsxtConfig.InsecureFlag,
+		RetriesConfiguration: retriesConfig,
+	}
+	client, err := nsxt.NewAPIClient(&cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating NSX-T client failed")
+	}
+	return NewNsxtBroker(client), nil
+}
+
+func (p *lbProvider) Initialize(client clientset.Interface, stop <-chan struct{}) {
+	go p.reorg(client.CoreV1().Services(""), stop)
+}
+
+// GetLoadBalancer returns the LoadBalancerStatus
+// Implementations must treat the *corev1.Service parameter as read-only and not modify it.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (p *lbProvider) GetLoadBalancer(ctx context.Context, clusterName string, service *corev1.Service) (status *corev1.LoadBalancerStatus, exists bool, err error) {
+	servers, err := p.access.FindVirtualServers(clusterName, objectNameFromService(service))
+	if err != nil {
+		return nil, false, err
+	}
+	if len(servers) == 0 {
+		return nil, false, nil
+	}
+	return newLoadBalancerStatus(servers[0].IpAddress), true, nil
+}
+
+func newLoadBalancerStatus(ipAddress string) *corev1.LoadBalancerStatus {
+	return &corev1.LoadBalancerStatus{
+		Ingress: []corev1.LoadBalancerIngress{
+			{IP: ipAddress},
+		},
+	}
+}
+
+// GetLoadBalancerName returns the name of the load balancer. Implementations must treat the
+// *corev1.Service parameter as read-only and not modify it.
+func (p *lbProvider) GetLoadBalancerName(ctx context.Context, clusterName string, service *corev1.Service) string {
+	return clusterName + ":" + service.Namespace + ":" + service.Name
+}
+
+// EnsureLoadBalancer creates a new load balancer 'name', or updates the existing one. Returns the status of the balancer
+// Implementations must treat the *corev1.Service and *corev1.Node
+// parameters as read-only and not modify them.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (p *lbProvider) EnsureLoadBalancer(ctx context.Context, clusterName string, service *corev1.Service, nodes []*corev1.Node) (*corev1.LoadBalancerStatus, error) {
+	key := objectNameFromService(service).String()
+	p.keyLock.Lock(key)
```

Comment:
Curious why the key-based lock. It seems like keyLock.Lock already holds a `sync.Mutex`, can we simplify and just lock `sync.Mutex` here?

Reply:
The Mutex in `keyLock` is used to lock the lock operation for a dedicated key. It is released again in the `keyLock.Lock` method. This locking is folded with the locking of a dedicated Mutex created for the given key to secure the map operation to hold the Mutex for the key.

Reply:
What benefit do we get from this instead of using sync.Mutex and locking against all keys? Is it for performance reasons so we can operate against multiple keys concurrently or something else?

Reply:
The dedicated lock per service might be useful to avoid the global reorg go routine to block all service reconcilations